crates/storage/src/address.rs

```diff
@@ -55,6 +58,34 @@ impl Id {
     }
 }
 
+impl BorshDeserialize for Id {
+    fn deserialize(buf: &mut &[u8]) -> Result<Self, IoError> {
+        if buf.len() < 16 {
+            return Err(IoError::new(
+                IoErrorKind::UnexpectedEof,
+                "Not enough bytes to deserialize Id",
+            ));
+        }
+        let (bytes, rest) = buf.split_at(16);
+        *buf = rest;
+        Ok(Self(Uuid::from_slice(bytes).map_err(|err| {
+            IoError::new(IoErrorKind::InvalidData, err)
+        })?))
+    }
```

Comment:
this can be omitted, the default implementation of deserialize will delegate to `deserialize_reader` which will return EOF since you're using `read_exact`

also, `Uuid::from_slice` does a length check internally

Reply:
Ah, that's good to know - I was unaware of that 🙂 I've not used Borsh before.